server/src/server.js

```diff
@@ -0,0 +1,156 @@
+app.use(express.json());
+
+app.use(
+  cookieSession({
+    name: "session",
+    keys: ["secret1"],
+    maxAge: 365 * 24 * 60 * 60 * 1000, // 1 year
+  })
+);
+
+app.use((req, res, next) => {
+  if (req.session.isNew) {
+    // Don't worry about collision
+    const uuid = uuidv4();
+    req.session.userID = uuid;
+  } else {
+    if (
+      !/^[a-z0-9]{8}-[a-z0-9]{4}-4[a-z0-9]{3}-[a-z0-9]{4}-[a-z0-9]{12}$/.test(
+        req.session.userID
+      )
+    ) {
+      return res.status(500).json({
+        status: 500,
+        error: { type: ERROR_BAD_UUID },
+      });
+    }
+  }
+  next();
+});
+
+/**
+ *  Middleware that can be used on routes that match `/:${URL_PARAM_EVENT_ID}/...`
+ *  This will fetch the event from Datastore and attach it and its key to
+ *  req.event and req.datastoreKey respectively.
+ *
+ *  It will return an error if, for any reason, the event could not be found.
+ */
+async function getEvent(req, res, next) {
+  const key = datastore.key([
+    KIND_EVENT,
+    // If the url is: /1/details
+    // then `req.params[URL_PARAM_EVENT_ID]` will be 1
+    parseInt(req.params[URL_PARAM_EVENT_ID]),
+  ]);
+  datastore
+    .runQuery(datastore.createQuery(KIND_EVENT).filter("__key__", "=", key))
+    .then((results) => {
+      // results[0] is the actual returned entities
+      // results[1] is metadata about the request (e.g are there additional results)
+      const events = results[0];
+      if (events.length === 0) {
+        res.status(400).json({
+          status: 400,
+          error: {
+            type: ERROR_INVALID_EVENT_ID,
+          },
+        });
+      } else {
+        const event = events[0];
+        if (
+          event === undefined ||
+          event === null ||
+          typeof event !== "object"
+        ) {
+          res.status(500).json({
+            status: 500,
+            error: {
+              type: ERROR_BAD_DB_INTERACTION,
+            },
+          });
+        } else {
+          req.event = event;
+          req.datastoreKey = key;
+          next();
+        }
+      }
+    })
+    .catch((_) => {
+      res.status(500).json({
+        status: 500,
+        error: {
+          type: ERROR_BAD_DB_INTERACTION,
+        },
+      });
+    });
+}
+
+app.post("/create", async (req, res) => {
+  const key = datastore.key([KIND_EVENT]);
+  const result = await datastore.save({ key, data: { users: {} } });
+  res.send({
+    // Datastore automatically generates a unique id
+    // for the key associated with our entity. This is the only
+    // way to get it :(
+    // TODO: Check if the production data store server uses
+    // random numbers. The local emulator increments the ids,
+    // which makes them easily guessable. That's bad!
+    eventID: result[0].mutationResults[0].key.path[0].id,
+  });
+});
+
+app.post(`/:${URL_PARAM_EVENT_ID}`, getEvent, async (req, res) => {
+  const { body, datastoreKey: key, event } = req;
```

Comment:
I am having some trouble understanding this section. Do you think you could explain to me the structure of the information being handled here? You are working with the request body, key, and event, but what exactly are those things? What information is inside of each and why?

Reply:
`body` is the request body. In the POST request to this server, the client must provide some data, which they will do by adding a body to the HTTP request that is a JSON string as follows:
```json
{
    "name": ved,
    "location": "[123, 321]"
}
```
The `express.json()` thing at the start of this file automatically parses bodies with the JSON content type into actual JS objects. Thus, `request.body` is a Javascript object.

Useful link: https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages

The request goes through some middleware before being processed by the handler function. Middleware can do various things to the request.

In this case, the request goes through the `getEvent` middleware. This middleware constructs a datastore key from the event id in the url, fetches the event from the database using the key, and then attaches both the key and the event to `req`.

Useful link: https://expressjs.com/en/guide/using-middleware.html

Reply:
Okay, I think I understand. Thank you for explaining that to me! So this post request is used to get event/session information from Datastore when updating it with new user information? And the get request (/:${URL_PARAM_EVENT_ID}/details) returns that event/session information when we want to send it back to the client?

Reply:
It matches the design docs. POST is for adding new user information. GET is for getting existing information. The getting event/session information is middleware that can be applied to either a POST or GET request.